src/modules/roc_pipeline/receiver_session_group.cpp

```diff
@@ -30,6 +30,8 @@ ReceiverSessionGroup::ReceiverSessionGroup(
     , mixer_(mixer)
     , receiver_state_(receiver_state)
     , receiver_config_(receiver_config) {
+    packet_ring_buffer_size_ = receiver_config_.common.max_sessions
+        * receiver_config_.default_session.max_session_packets;
```

Comment:
I have an idea.

Currently we're pre-defining the maximum number of packets, and actually forcing the user to specify it (because any default value will be applicable only for some configurations, but wont work for others).

Instead, can use the fact that packets in the ring buffer are actually sorted by the arrival time, and limit the buffer size by the age of the oldest received packet? In other words, we'll remove packets from the head until the first packet is older than N seconds.

In this case, we can limit the buffer size to the maximum allowed latency (maybe with a factor of 1.5 or so), which we already have in our config. It would be a good limit and we wont need max_sessions and max_session_packets at all.

If needed, we can add the arrival time field to the packet. We can fill it in roc_netio. Take a look at core::timestamp() function.

Reply:
cool, I see there is `timestamp_t begin() const;` and `timestamp_t end() const;` for the timestamp of first sample and last sample of each packet, is it similar to the arrival time of packet?

Reply:
No, this values are based on packet timestamp field from RTP, which is the logical position of the packet in the stream. And here we need a timestamp from the wall clock.

Reply:
it this the location where packet is constructed? https://github.com/roc-project/roc/blob/develop/src/modules/roc_netio/target_libuv/roc_netio/udp_receiver_port.cpp#L273-L287

Reply:
Yes, I think this is a good place to fill the arrival time.